crates/ruff_diagnostics/src/fix.rs

```diff
@@ -4,23 +4,106 @@ use serde::{Deserialize, Serialize};
 
 use crate::edit::Edit;
 
+/// Indicates confidence in the correctness of a suggested fix.
+#[derive(Default, Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[non_exhaustive]
+pub enum Applicability {
+    /// The fix is definitely what the user intended, or maintains the exact meaning of the code.
+    /// This fix should be automatically applied.
+    Automatic,
+
+    /// The fix may be what the user intended, but it is uncertain.
+    /// The fix should result in valid code if it is applied.
+    /// The fix can be applied with user opt-in.
+    Suggested,
+
+    /// The fix has a good chance of being incorrect or the code be incomplete.
+    /// The fix may result in invalid code if it is applied.
+    /// The fix should only be manually applied by the user.
+    Manual,
+
+    /// The applicability of the fix is unknown.
+    #[default]
+    Unspecified,
+}
+
 /// A collection of [`Edit`] elements to be applied to a source file.
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct Fix {
     edits: Vec<Edit>,
+    applicability: Applicability,
 }
 
 impl Fix {
     /// Create a new [`Fix`] with an unspecified applicability from an [`Edit`] element.
+    #[deprecated(
+        note = "Use `Fix::automatic`, `Fix::suggested`, or `Fix::manual` instead to specify an applicability."
+    )]
     pub fn unspecified(edit: Edit) -> Self {
-        Self { edits: vec![edit] }
+        Self {
+            edits: vec![edit],
+            applicability: Applicability::Unspecified,
+        }
     }
 
-    /// Create a new [`Fix`] with unspecified applicability from multiple [`Edit`] elements.
+    /// Create a new [`Fix`] with an unspecified applicability from multiple [`Edit`] elements.
+    #[deprecated(
+        note = "Use `Fix::automatic_edits`, `Fix::suggested_edits`, or `Fix::manual_edits` instead to specify an applicability."
+    )]
     pub fn unspecified_edits(edit: Edit, rest: impl IntoIterator<Item = Edit>) -> Self {
         Self {
             edits: std::iter::once(edit).chain(rest.into_iter()).collect(),
+            applicability: Applicability::Unspecified,
+        }
+    }
+
+    /// Create a new [`Fix`] with automatic applicability from an [`Edit`] element.
```

Comment:
I recommend repeating the `automatic` applicability documentation here. Most users will only use the `Fix` API, and explaining what `automatic` safes them a jump in the documentation (or make sure they read it).

```suggestion
    /// Create a new [`Fix`] with [automatic applicability](Applicability::Automatic) from an [`Edit`] element.
```

Reply:
👍 did this for all of the levels